core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -220,21 +257,53 @@ Iterable<Partition> all() {
 
   static class Partition {
     private final StructLike key;
-    private long recordCount;
-    private int fileCount;
     private int specId;
+    private long dataRecordCount;
+    private int dataFileCount;
+
+    private final Set<DeleteFile> equalityDeleteFiles;
+    private final Set<DeleteFile> positionDeleteFiles;
 
     Partition(StructLike key) {
       this.key = key;
-      this.recordCount = 0;
-      this.fileCount = 0;
       this.specId = 0;
+      this.dataRecordCount = 0;
+      this.dataFileCount = 0;
+      this.positionDeleteFiles = Sets.newHashSet();
+      this.equalityDeleteFiles = Sets.newHashSet();
+    }
+
+    private void update(FileScanTask task) {
```

Comment:
I think I would strongly want to consider an approach which either disposes of the set's when the partition info is done being constructed or doesn't use this set approach. It looks like in the current implementation we end up keeping the entire set of delete file objects in memory indefinitely.

Reply:
I also had the similar concern, and thought of using PositionDeletesTableScan, but it seems there is no equivalent EqualityDeletesTableScan yet.  

Maybe another approach here could be, going through each partition and using DeleteFileIndex one by one.

Reply:
Actually that way I thought will be quite expensive (two pass).  

Probably the only way to effectively do it , until this whole table is migrated over to some kind of view of 'files' table, is to rewrite the PartitionsTableScan to directly use the underlying code:  ManifestReader.readDeleteManifest() / ManifestReader.read(), and then go through it, instead of using the ManifestGroup.planFiles() way.  

That way, we can iterate through , and collect delete files/data files in one pass.  Any thoughts?

Reply:
> Probably the only way to effectively do it , until this whole table is migrated over to some kind of view of 'files' table, is to rewrite the PartitionsTableScan to directly use the underlying code: ManifestReader.readDeleteManifest() / ManifestReader.read(), and then go through those iterators, instead of using the ManifestGroup.planFiles() / FileScanTask way.

@szehon-ho: I have spent some time and realized that I am not super familiar with this side of code.  Would you like to contribute a PR for data files for this? I can then extend it to delete files and handle these stats updates.

Reply:
Yea, I am thinking something along the lines of what currently BaseFilesTable/ ManifestsTable does, sure let me take a look , when I get a chance.

Reply:
Thanks. I will also explore those files.

BTW I have updated to Map<String, Integer> instead of set<DataFile> and cleared it after the usage now. But looking forward to solving it as you suggested by changing the `planFiles`